da/da.go

```diff
@@ -236,9 +238,174 @@ func (dac *DAClient) RetrieveHeaders(ctx context.Context, dataLayerHeight uint64
 	}
 }
 
-func (dac *DAClient) submit(ctx context.Context, blobs []goDA.Blob, gasPrice float64, namespace goDA.Namespace) ([]goDA.ID, error) {
+func (dac *DAClient) submit(ctx context.Context, blobs []coreda.Blob, gasPrice float64, namespace coreda.Namespace) ([]coreda.ID, error) {
 	if len(dac.SubmitOptions) == 0 {
 		return dac.DA.Submit(ctx, blobs, gasPrice, namespace)
 	}
 	return dac.DA.SubmitWithOptions(ctx, blobs, gasPrice, namespace, dac.SubmitOptions)
 }
+
+//--------------------------------
+// Batches
+//--------------------------------
+
+// ResultSubmitBatch contains information returned from DA layer after block headers/data submission.
+type ResultSubmitBatch struct {
+	BaseResult
+	// Not sure if this needs to be bubbled up to other
+	// parts of Rollkit.
+	// Hash hash.Hash
+}
+
+// ResultRetrieveBatch contains batch of block data returned from DA layer client.
+type ResultRetrieveBatch struct {
+	BaseResult
+	// Data is the block data retrieved from Data Availability Layer.
+	// If Code is not equal to StatusSuccess, it has to be nil.
+	Data []*types.Data
+}
+
+// SubmitBatch submits block data to DA.
+func (dac *DAClient) SubmitBatch(ctx context.Context, data []*coresequencer.Batch, maxBlobSize uint64, gasPrice float64) ResultSubmitBatch {
+	var (
+		blobs    [][]byte
+		blobSize uint64
+		message  string
+	)
+	for i := range data {
+		protoBatch := &pb.Batch{
+			Txs: data[i].Transactions,
+		}
+		blob, err := proto.Marshal(protoBatch)
+		if err != nil {
+			message = fmt.Sprint("failed to serialize block", err)
+			dac.logger.Info(message)
+			break
+		}
+		if blobSize+uint64(len(blob)) > maxBlobSize {
+			message = fmt.Sprint((&ErrBlobSizeOverLimit{}).Error(), "blob size limit reached", "maxBlobSize", maxBlobSize, "index", i, "blobSize", blobSize, "len(blob)", len(blob))
+			dac.logger.Info(message)
+			break
+		}
+		blobSize += uint64(len(blob))
+		blobs = append(blobs, blob)
+	}
+	if len(blobs) == 0 {
+		return ResultSubmitBatch{
+			BaseResult: BaseResult{
+				Code:    StatusError,
+				Message: "failed to submit blocks: no blobs generated " + message,
+			},
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, dac.SubmitTimeout)
+	defer cancel()
+	ids, err := dac.DA.Submit(ctx, blobs, gasPrice, dac.Namespace)
+	if err != nil {
+		status := StatusError
+		switch {
+		case errors.Is(err, &ErrTxTimedOut{}):
+			status = StatusNotIncludedInBlock
+		case errors.Is(err, &ErrTxAlreadyInMempool{}):
+			status = StatusAlreadyInMempool
+		case errors.Is(err, &ErrTxIncorrectAccountSequence{}):
+			status = StatusAlreadyInMempool
+		case errors.Is(err, &ErrTxTooLarge{}):
+			status = StatusTooBig
+		case errors.Is(err, &ErrContextDeadline{}):
+			status = StatusContextDeadline
+		}
+		return ResultSubmitBatch{
+			BaseResult: BaseResult{
+				Code:    status,
+				Message: "failed to submit block data: " + err.Error(),
+			},
+		}
+	}
+
+	if len(ids) == 0 {
```

Comment:
What if number of ids is different than number of blobs (but not zero)?

Reply:
@gupadhyaya can you help answer this?

Reply:
we can add those checks. this was the code moved from rollkit originally.

Reply:
added